api/src/wfl/module/covid.clj

```diff
@@ -23,6 +24,61 @@
 (defn create-covid-workload!
   [tx {:keys [source sink executor] :as request}])
 
+(defn start-covid-workload!
+  [{:keys [submission-id workspace] :as workload}])
+
+(defn ^:private snapshot-error-map
+  "Return default ExceptionInfo map for snapshot-related exceptions."
+  [{:keys [items sink] :as _workload} {:keys [id snapshot_id] :as _record}]
+  {:workspace (:workspace sink)
+   :table items
+   :id id
+   :snapshot_id snapshot_id})
+
+(defn ^:private try-get-snapshot-reference
+  "Attempt to fetch snapshot reference with id SNAPSHOT_REFERENCE_ID
+  from SINK-derived workspace."
+  [{:keys [sink] :as workload} {:keys [snapshot_reference_id] :as record}]
+  (let [workspace (:workspace sink)]
+    (try
+      (rawls/get-snapshot-reference workspace snapshot_reference_id)
+      (catch Throwable cause
+        (throw (ex-info "Rawls unable to fetch snapshot reference"
+                        (util/deep-merge
+                         (snapshot-error-map workload record)
+                         {:snapshot_reference_id snapshot_reference_id})
+                        cause))))))
+
+(defn ^:private snapshot-imported?
+  "Check if snapshot with id SNAPSHOT_ID in TDR has been imported to
+  SINK-derived workspace and can be successfully fetched."
+  [workload {:keys [snapshot_id snapshot_reference_id] :as record}]
+  {:pre [(some? snapshot_id)]}
+  (and (some? snapshot_reference_id)
+       (some? (try-get-snapshot-reference workload record))))
+
+(defn ^:private import-snapshot!
+  "Import snapshot with id SNAPSHOT_ID from TDR to workspace.
+  Use transaction TX to update ITEMS table with resulting reference id."
+  [tx {:keys [items sink] :as workload} {:keys [id snapshot_id] :as record}]
+  (let [workspace (:workspace sink)
+        name (util/randomize "placeholder-snapshot-ref-name")
```

Comment:
As we know, naming things is hard…

This will be the name of the snapshot reference in the workspace.  It needs to be unique, but AFAIK is only rendered in the UI.  We never pass it to Rawls.

- Any preferences for what this name should be?
- Do we want to punt name generation to `rawls/create-snapshot-reference`?

Reply:
Can we use the name of the snapshot instead of randomising?

Reply:
Rawls permits us to create multiple snapshot references for the same source snapshot as long as the snapshot reference names are unique.

If we want to allow for this possibility, we can use the snapshot name as the base name with a random suffix.

Reply:
Here's an example snapshot created in this manner, where the source snapshot has name "sarscov2_illumina_full_inputs":
```
{:cloningInstructions "COPY_NOTHING",
 :description "",
 :name "sarscov2_illumina_full_inputs78aa97140e6f402dae0ecba77a563a01",
 :reference {:instanceName "terra", :snapshot "7cb392d8-949b-419d-b40b-d039617d2fc7"},
 :referenceId "f7b7eb31-cdc8-4eb4-a42b-18355dc716ee",
 :referenceType "DATA_REPO_SNAPSHOT",
 :workspaceId "e9d053b9-d79f-40b7-b701-904bf542ec2d"}
```